### PR TITLE
feat: [MR-552] Implement callback expiration 

### PR DESCRIPTION
**Pull request by alin-at-dfinity**
[MR-552]: Generate compact reject responses upon best-effort callback expiration. And inflate them into SYS_UNKNOWN reject responses when peeking/popping.

[MR-552]: https://dfinity.atlassian.net/browse/MR-552?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: rs/replicated_state/src/canister_state/queues/message_pool/tests.rs
Comment (stiegerc) [inline rs/replicated_state/src/canister_state/queues/message_pool/tests.rs:159]: I guess it's a matter of taste but I usually define the function before I use it in cases like this.

File: rs/replicated_state/src/canister_state/queues/message_pool/tests.rs
Comment (stiegerc) [inline rs/replicated_state/src/canister_state/queues/message_pool/tests.rs:418]: Personal preference: you write this using `assert_matches` without `unwrap().0`. And other places too.

**Review by stiegerc — COMMENTED** [review]
So far lgtm. Didn't spend much time on the tests though.

File: rs/replicated_state/src/canister_state/queues.rs
Comment (alin-at-dfinity) [reply rs/replicated_state/src/canister_state/queues.rs]: It's only called by `test_invariants()` (often, but not in production code) and while converting from a proto.

The reason I didn't make it a `BTreeSet` directly is that I want to detect duplicates, by comparing the size of the `Vec` to the size of the `BTreeSet` (see just below). I had the approach you suggest at some point, but it looked kind of awkward. I went with it again. It still looks a bit awkward, but at least it doesn't also allocate a `Vec` on the side.

File: rs/replicated_state/src/canister_state/queues/message_pool/tests.rs
Comment (alin-at-dfinity) [reply rs/replicated_state/src/canister_state/queues/message_pool/tests.rs]: It works, but it's more verbose:
```rust
    assert_matches!(
        pool.shed_largest_message(),
        Some((SomeReference::Inbound(r), _)) if r == ref2
    );
```

File: rs/replicated_state/src/canister_state/queues/message_pool/tests.rs
Comment (alin-at-dfinity) [reply rs/replicated_state/src/canister_state/queues/message_pool/tests.rs]: I did it this way to minimize the diff. But you're right, it is probably not worth the confusion.

**Review by derlerd-dfinity — COMMENTED** [review]
Did a quick initial pass and left some comments.

File: rs/protobuf/def/state/queues/v1/queues.proto
Comment (alin-at-dfinity) [reply rs/protobuf/def/state/queues/v1/queues.proto:205]: Yes, it is. It was added later (hence the tag number), but in the Rust code (and conceptually, in terms of importance) expired callbacks come before shed responses.

Looking at it from the opposite perspective, I really-really don't like proto message definitions that are just a list of fields in the order in which they were added, with no attention given to structure / logical grouping.

File: rs/replicated_state/src/canister_state/queues.rs
Comment (alin-at-dfinity) [reply rs/replicated_state/src/canister_state/queues.rs]: I switched it around to say `pool.get(_).is_none() && !(BestEffort && Response && (expired_callbacks.contains_key(_) || shed_responses.contains_key(_)))`. Hopefully that's clearer.

The alternative is to either write it like:
```rust
if self.pool.get(reference).is_some() {
    return false;
}

if reference.class == Class::BestEffort && reference.kind() == Kind::Response {
    self.expired_callbacks.contains_key(&reference)
        || self.shed_responses.contains_key(&reference)
} else {
    true
}
```
But that doesn't look more readable to me.

Or, we could do something like:
```rust
let is_in_pool = self.pool.get(reference).is_some();
let is_best_effort_response =
    reference.class == Class::BestEffort && reference.kind() == Kind::Response;
let is_expired_callback = self.expired_callbacks.contains_key(&reference);
let is_shed_response = self.shed_responses.contains_key(&reference);

return is_in_pool
    || is_best_effort_response && (is_expired_callback || is_shed_response);
```

But this would be both inefficient (every lookup is made every single time) and, again, not particularly readable either.

Comment (derlerd-dfinity) [reply rs/replicated_state/src/canister_state/queues.rs]: I think the option you ended up with makes sense. Thank you.

File: rs/replicated_state/src/canister_state/queues.rs
Comment (stiegerc) [reply rs/replicated_state/src/canister_state/queues.rs]: Fair enough. How about a function
```Rust
fn is_best_effort_response(&self) -> bool {
   self.class() == Class::BestEffort && self.kind() == Kind::Response
}
```
or something like that? Maybe also something similar for `expired_callbacks` and `shed_responses` since both of these expressions also appear in the complicated logical expression @derlerd-dfinity mentioned below?

File: rs/replicated_state/src/canister_state/queues/message_pool/tests.rs
Comment (stiegerc) [reply rs/replicated_state/src/canister_state/queues/message_pool/tests.rs]: I was hoping it would work without an `if` clause. I tend to prefer a triggering `assert` rather than an `unwrap()` inside of it but feel free to leave it as it is.

File: rs/replicated_state/src/canister_state/queues.rs
Comment (stiegerc) [reply rs/replicated_state/src/canister_state/queues.rs]: > The goal of mentioning an approximate time complexity is to make it clear that these are debug- or loading-time only checks

I see, I didn't realize that.

File: rs/replicated_state/src/canister_state/queues.rs
Comment (derlerd-dfinity) [inline rs/replicated_state/src/canister_state/queues.rs:567]: The name `reserve_inbound_timeout_response` makes me think of slot reservations in queues, while we are just reserving an id (resp. reference) here. Maybe consider renaming.

File: rs/replicated_state/src/canister_state/system_state.rs
Comment (derlerd-dfinity) [inline rs/replicated_state/src/canister_state/system_state.rs:1484]: Is this unwrap here really safe? In particular, could there be cases where we interact with a dishonest subnet and one of the error conditions is triggered?

Even if it is safe, it might be worth thinking about raising a critical error, logging all the information and going ahead. It would seem to me that there is not much we win if we crash because we can't induct a timeout response.

File: rs/replicated_state/src/canister_state/queues/tests.rs
Comment (oggy-dfin) [inline rs/replicated_state/src/canister_state/queues/tests.rs:78]: Absolutely not for this PR, but conceptually there should be one callback ID for every successful `push_output_request`. If you're going in the direction of bundling the `CallContextManager` together with the queues, I do wonder if we have any practical chance of actually enforcing this, i.e., somehow have `CallbackId` be only generated through `push_output_request`. I realize you'd need to thread some canister-global state through to keep the numbers unique across the different queues.  And I now also realize that there are two `next_callback_id` fields, one in `SandboxSafeSystemState`, and one in `CallContextManager`, and I have no idea what's going on there.

File: rs/replicated_state/src/canister_state/system_state.rs
Comment (oggy-dfin) [reply rs/replicated_state/src/canister_state/system_state.rs:1484]: Same question, wouldn't a malicious sender send a response for an nonexistent request, for example?

File: rs/replicated_state/src/canister_state/system_state/call_context_manager.rs
Comment (oggy-dfin) [inline rs/replicated_state/src/canister_state/system_state/call_context_manager.rs:279]: Nit, but I found the previous formulation more informative, it's not really clear to me what `n` is if I just see that.

File: rs/replicated_state/tests/system_state.rs
Comment (oggy-dfin) [inline rs/replicated_state/tests/system_state.rs:165]:
```suggestion
    /// Times out all callbacks with deadlines before `current_time`. Returns the
```

**Review by oggy-dfin — COMMENTED** [review]
Thanks! LGTM overall, the blocker for me is understanding why the unwrap in system_state is safe (assuming that it is safe).

File: rs/replicated_state/src/canister_state/queues/message_pool.rs
Comment (derlerd-dfinity) [inline rs/replicated_state/src/canister_state/queues/message_pool.rs:122]: Nice! I like that this is encapsulated now as it makes the places in the code where this is checked much easier to read.

Do we need (or maybe even already have) some tests around these functions? It would appear that they are easy to break in subtle ways.

File: rs/replicated_state/src/canister_state/queues/message_pool.rs
Comment (derlerd-dfinity) [inline rs/replicated_state/src/canister_state/queues/message_pool.rs:161]: Why don't we need class at all anymore?

File: rs/replicated_state/src/canister_state/system_state.rs
Comment (derlerd-dfinity) [inline rs/replicated_state/src/canister_state/system_state.rs:1464]: Do we really still want to return a state error here? If we'd introduce some custom error type that can accumulate the failed callbacks we could even log some information that might be relevant for a manual recovery in such a case.

File: rs/replicated_state/src/canister_state/system_state.rs
Comment (derlerd-dfinity) [inline rs/replicated_state/src/canister_state/system_state.rs:1777]: Do we need any warning related to not making this public, or should we maybe even inline it given that it is only used once?

**Review by derlerd-dfinity — COMMENTED** [review]
Just finished my pass over the prod code. Left some more questions comments but overall LGTM. 

Will complete my pass by going over the tests; after that I think I should be ready to approve.

File: rs/replicated_state/src/canister_state/system_state.rs
Comment (alin-at-dfinity) [reply rs/replicated_state/src/canister_state/system_state.rs]: I could return a `Vec<StateError>`, if we're concerned about failing to enqueue multiple callbacks and want to rely on the actual error message to recover them. But I believe the details in `StateError::NonMatchingResponse`(why we failed to enqueue it, the respondent, the deadline) are actually useful for debugging the root cause (is this a subnet-local message? was the queue already gone? etc.).

File: rs/replicated_state/src/canister_state/system_state.rs
Comment (alin-at-dfinity) [reply rs/replicated_state/src/canister_state/system_state.rs:1782]: This same function comes in the "encapsulate `CallContextManager`" PR. Likely with a warning attached. But that is unnecessary, technically speaking, since the `CallContextManager::status` will become private anyway.

File: rs/replicated_state/tests/system_state.rs
Comment (derlerd-dfinity) [inline rs/replicated_state/tests/system_state.rs:491]: Should we peek the input here to see that before the timeout call the message is still untouched?

**Review by derlerd-dfinity — APPROVED** [review]
Thanks a lot. LGTM as soon as Ogi and Christian are happy as well.

File: rs/replicated_state/tests/system_state.rs
Comment (alin-at-dfinity) [reply rs/replicated_state/tests/system_state.rs:491]: Before the `time_out_callbacks()` call below, there are no messages, only the callback. That's what the `assert_eq!(None, pop_input())` above checks.

Comment (derlerd-dfinity) [reply rs/replicated_state/tests/system_state.rs:491]: Of course -- sorry, seems I was somehow confused. Thanks for clarifying.

File: rs/replicated_state/src/canister_state/system_state.rs
Comment (oggy-dfin) [inline rs/replicated_state/src/canister_state/system_state.rs:1482]: I now realize I hadn't thought about DTS earlier. Are we sure we'll never push a late response into an input queue if we previously pushed a timeout response, and that timeout response happens to trigger a DTS execution that's progressing exactly as the "real" response comes in?

Also, if a DTS execution gets aborted (e.g., because of a checkpoint), do we need to reinsert it into the `unexpired_callbacks` again?

File: rs/replicated_state/src/canister_state/system_state/call_context_manager.rs
Comment (oggy-dfin) [reply rs/replicated_state/src/canister_state/system_state/call_context_manager.rs:279]: I'd then rather write "EXPENSIVE - DO NOT CALL REPEATEDLY" or something along those lines. But I leave it up to you.

**Review by oggy-dfin — COMMENTED** [review]
I have two a couple questions about the interplay with DTS that I missed in the first round - hopefully noops.

File: rs/replicated_state/src/canister_state/system_state.rs
Comment (alin-at-dfinity) [reply rs/replicated_state/src/canister_state/system_state.rs:1482]: Hmm... You may have uncovered one very nasty bug.

We don't explicitly check against a possible aborted or paused response when inducting responses. I guess I was assuming that the respective `CallbackId` will stay in `CanisterQueues::callbacks_with_enqueued_response` until the callback is unregistered. But that's not the case: we remove it from that set as soon as the response is popped.

Regardless, even if we did explicitly check against na aborted or paused response execution, there would still be a loophole. While that would prevent inducting a response "duplicating" the already paused or aborted one, it would not prevent us from producing a reject response by timing out an outbound request. While I don't think that's an actual issue (because we first time out messages, then callbacks; so we would first produce an actual reject response, then try to produce a `DeadlibeExpired` compact response here), that's still not something I'm comfortable relying on.

But the somewhat hacky, yet safer way around that would be to remember the last `CallbackId` we popped; and include that alongside `CanisterQueues::callbacks_with_enqueued_response` in the duplicate response check.

I'll think about this some more, since it's such a nasty one. And thank you for asking the question.

**Pull request merged**
